File: Practica2/busquedaBinaria.py
import tracemalloc
import time


sizeArr = int(input("Ingresa el tamaño del arreglo que deseas: ")) + 1
numeroBuscar = int(input("Ingresa el numero que deseas buscar: "))

arr = list(range(1, sizeArr))

start = time.time()
tracemalloc.start()

arr.sort()

def busquedaBinaria(arr,numeroBuscar,inicio,final): 
    
    if (inicio > final):
        return -1
    # size = final - 1
    # print(size)
    mid = (inicio + final) // 2
    
    # print(f"idx {mid} value {arr[mid]}")
    # print(f"Inicio {inicio} Final {final} Mid {mid}")

    if(arr[mid] == numeroBuscar):
        return arr[mid]
    if(numeroBuscar < arr[mid]):
        return busquedaBinaria(arr,numeroBuscar,inicio,mid-1)
    if(numeroBuscar > arr[mid]):
        return busquedaBinaria(arr,numeroBuscar,mid+1,final)
    if(numeroBuscar > final or numeroBuscar < inicio):
        return -1
 

# def busquedaBinaria(arr,numeroBuscar):
size = len(arr)
    

posicion = busquedaBinaria(arr, numeroBuscar,0, size)
    
if(posicion != -1):
    print(f"Numero encontrado en la posicion {posicion}")
else:
    print(f"Numero no encontrado")



current, peak = tracemalloc.get_traced_memory()
print(f"Memoria actual: {current / 1024} KB; Pico: {peak / 1024} KB")
tracemalloc.stop()
end = time.time()
print(f"Tiempo de ejecución: {end - start} segundos")

File: Practica2/Busqueda Binaria.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int busquedaBinaria(int arr[], int n, int x) {
    int inicio = 0, fin = n - 1;
    while (inicio <= fin) {
        int medio = (inicio + fin) / 2;
        if (arr[medio] == x) return medio;
        else if (arr[medio] < x) inicio = medio + 1;
        else fin = medio - 1;
    }
    return -1;
}

int main() {
    int n, x;
    printf("Ingrese el tamaño del arreglo: ");
    scanf("%d", &n);

    int *arr = (int*)malloc(n * sizeof(int));
    if (arr == NULL) return 1;

    for (int i = 0; i < n; i++) arr[i] = i + 1;

    printf("Ingrese el valor a buscar: ");
    scanf("%d", &x);

    clock_t inicio = clock();
    int resultado = busquedaBinaria(arr, n, x);
    clock_t fin = clock();

    double tiempo = (double)(fin - inicio) / CLOCKS_PER_SEC;
    size_t memoria = n * sizeof(int);

    if (resultado == -1)
        printf("No encontrado\n");
    else
        printf("Encontrado en la posición %d\n", resultado);

    printf("Tiempo: %f segundos\n", tiempo);
    printf("Memoria usada: %zu bytes\n", memoria);

    free(arr);
    return 0;
}